### PR TITLE
VDS-689: Add cloning of product parts

### PR DESCRIPTION
**Pull request by asvishnyakov**

### CODE REVIEW

File: src/VirtoCommerce.DemoSolutionFeaturesModule.Core/Models/Catalog/DemoProductPart.cs [hunk 1]
```diff
@@ -24,5 +26,20 @@ public class DemoProductPart : AuditableEntity, IHasName
         public string DefaultItemId { get; set; }
 
         public ProductPartItemInfo[] PartItems { get; set; }
+
+        public virtual object Clone()
+        {
+            var result = (DemoProductPart)MemberwiseClone();
```
Comment (pushnitsa) [inline src/VirtoCommerce.DemoSolutionFeaturesModule.Core/Models/Catalog/DemoProductPart.cs]: Add space after casting

File: src/VirtoCommerce.DemoSolutionFeaturesModule.Core/Models/Catalog/DemoProduct.cs [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+namespace VirtoCommerce.DemoSolutionFeaturesModule.Core.Models.Catalog
+{
+    public class DemoProduct: CatalogProduct
+    {
+        public DemoProductPart[] ProductParts { get; set; }
+
+        public override object GetCopy()
+        {
+            var result = (DemoProduct)base.GetCopy();
```
Comment (pushnitsa) [inline src/VirtoCommerce.DemoSolutionFeaturesModule.Core/Models/Catalog/DemoProduct.cs]: Add space after casting

File: src/VirtoCommerce.DemoSolutionFeaturesModule.Core/Models/Catalog/DemoProduct.cs [hunk 3]
```diff
@@ -0,0 +1,17 @@
+using System.Linq;
+using VirtoCommerce.CatalogModule.Core.Model;
+
+namespace VirtoCommerce.DemoSolutionFeaturesModule.Core.Models.Catalog
+{
+    public class DemoProduct: CatalogProduct
+    {
+        public DemoProductPart[] ProductParts { get; set; }
```
Comment (pushnitsa) [inline src/VirtoCommerce.DemoSolutionFeaturesModule.Core/Models/Catalog/DemoProduct.cs:8]: Initialize empty array in the constructor

File: src/VirtoCommerce.DemoSolutionFeaturesModule.Core/Models/Catalog/DemoProductPartItemInfo.cs [hunk 4]
```diff
@@ -1,10 +1,17 @@
...
-    public class ProductPartItemInfo
+    public class ProductPartItemInfo: ICloneable
     {
         public string ItemId { get; set; }
 
         public int Priority { get; set; }
+
+        public object Clone()
+        {
+            return (ProductPartItemInfo)MemberwiseClone();
```
Comment (pushnitsa) [inline src/VirtoCommerce.DemoSolutionFeaturesModule.Core/Models/Catalog/DemoProductPartItemInfo.cs]: Add space after casting

File: src/VirtoCommerce.DemoSolutionFeaturesModule.Web/Controllers/Api/DemoCatalogController.cs [hunk 5]
```diff
@@ -13,12 +13,12 @@ namespace VirtoCommerce.DemoSolutionFeaturesModule.Web.Controllers.Api
     [Route("api/demo/catalog")]
     public class DemoCatalogController : Controller
     {
-        private readonly IDemoProductPartSerarchService _partsSerarchService;
+        private readonly IDemoProductPartSearchService _partsSerarchService;
```
Comment (pushnitsa) [inline src/VirtoCommerce.DemoSolutionFeaturesModule.Web/Controllers/Api/DemoCatalogController.cs]: Please rename also field name

File: src/VirtoCommerce.DemoSolutionFeaturesModule.Data/Repositories/Catalog/DemoCatalogRepository.cs [hunk 6]
```diff
@@ -17,13 +19,24 @@ public DemoCatalogRepository(DemoCatalogDbContext dbContext)
         {
         }
 
+        public override async Task<ItemEntity[]> GetItemByIdsAsync(string[] itemIds, string responseGroup = null)
+        {
+            var result = await base.GetItemByIdsAsync(itemIds, responseGroup);
+            var itemResponseGroup = EnumUtility.SafeParseFlags(responseGroup, ItemResponseGroup.ItemLarge);
+            if (!itemIds.IsNullOrEmpty() && result.Any() && itemResponseGroup.HasFlag(ItemResponseGroup.ItemLarge))
```
Comment (pushnitsa) [inline src/VirtoCommerce.DemoSolutionFeaturesModule.Data/Repositories/Catalog/DemoCatalogRepository.cs:27]: Add empty string before and after if statement

File: src/VirtoCommerce.DemoSolutionFeaturesModule.Data/Models/Catalog/DemoItemEntity.cs [hunk 7]
```diff
@@ -1,11 +1,45 @@
...
+        {
+            var demoProduct = (DemoProduct) base.ToModel(product, convertChildrens, convertAssociations);
+            demoProduct.ProductParts = ConfiguredProductParts.Select(x =>
+                x.ToModel(AbstractTypeFactory<DemoProductPart>.TryCreateInstance())).ToArray();
+            return demoProduct;
+        }
+
+        public override ItemEntity FromModel(CatalogProduct product, PrimaryKeyResolvingMap pkMap)
+        {
+            var demoProduct = (DemoProduct) product;
```
Comment (pushnitsa) [inline src/VirtoCommerce.DemoSolutionFeaturesModule.Data/Models/Catalog/DemoItemEntity.cs]: if (product is DemoProduct demoProduct)

File: src/VirtoCommerce.DemoSolutionFeaturesModule.Data/Models/Catalog/DemoItemEntity.cs [hunk 8]
```diff
@@ -1,11 +1,45 @@
...
+                    AbstractTypeFactory<DemoProductPartEntity>.TryCreateInstance().FromModel(x, pkMap)));
+            }
+            return demoItemEntity;
+        }
+
+        public override void Patch(ItemEntity target)
+        {
+            base.Patch(target);
+
+            var demoItemEntity = (DemoItemEntity) target;
```
Comment (pushnitsa) [inline src/VirtoCommerce.DemoSolutionFeaturesModule.Data/Models/Catalog/DemoItemEntity.cs]: if (target is DemoItemEntity demoItemEntity)

File: src/VirtoCommerce.DemoSolutionFeaturesModule.Data/Models/Catalog/DemoItemEntity.cs [hunk 9]
```diff
@@ -1,11 +1,45 @@
...
 
 namespace VirtoCommerce.DemoSolutionFeaturesModule.Data.Models.Catalog
 {
     public class DemoItemEntity : ItemEntity
     {
         public virtual ObservableCollection<DemoProductPartEntity> ConfiguredProductParts { get; set; } = new NullCollection<DemoProductPartEntity>();
+
+        public override CatalogProduct ToModel(CatalogProduct product, bool convertChildrens = true, bool convertAssociations = true)
+        {
+            var demoProduct = (DemoProduct) base.ToModel(product, convertChildrens, convertAssociations);
```
Comment (pushnitsa) [inline src/VirtoCommerce.DemoSolutionFeaturesModule.Data/Models/Catalog/DemoItemEntity.cs]: var product = base.ToModel(product, convertChildrens, convertAssociations);

if (product is DemoProduct demoProduct)

**Review by pushnitsa — CHANGES_REQUESTED** [review]

**Review by Vectorfield4 — CHANGES_REQUESTED** [review]
Can you add a test to the `GetItemByIdsAsync` method for `DemoCatalogRepository`?

File: src/VirtoCommerce.DemoSolutionFeaturesModule.Core/Models/Catalog/DemoProduct.cs (hunk 3 above)
Comment (asvishnyakov) [reply src/VirtoCommerce.DemoSolutionFeaturesModule.Core/Models/Catalog/DemoProduct.cs:8]: We should initialize domain model arrays in constructors, we talked about that already few times. Null here is valid valuel; see how it's processed in data models

File: src/VirtoCommerce.DemoSolutionFeaturesModule.Data/Models/Catalog/DemoItemEntity.cs (hunk 9 above)
Comment (asvishnyakov) [reply src/VirtoCommerce.DemoSolutionFeaturesModule.Data/Models/Catalog/DemoItemEntity.cs]: We override product type so all instances always MUST be of our type. If the system somehow will create an instance of the base type, then it already means it's in an inconsistent state. So I think we shouldn't add additional checks here, just explicitly cast to our type.

File: src/VirtoCommerce.DemoSolutionFeaturesModule.Data/Repositories/Catalog/DemoCatalogRepository.cs (hunk 6 above)
Comment (asvishnyakov) [reply src/VirtoCommerce.DemoSolutionFeaturesModule.Data/Repositories/Catalog/DemoCatalogRepository.cs:27]: Fixed

File: src/VirtoCommerce.DemoSolutionFeaturesModule.Core/Models/Catalog/DemoProductPart.cs (hunk 1 above)
Comment (asvishnyakov) [reply src/VirtoCommerce.DemoSolutionFeaturesModule.Core/Models/Catalog/DemoProductPart.cs]: Fixed

File: src/VirtoCommerce.DemoSolutionFeaturesModule.Core/Models/Catalog/DemoProduct.cs (hunk 2 above)
Comment (asvishnyakov) [reply src/VirtoCommerce.DemoSolutionFeaturesModule.Core/Models/Catalog/DemoProduct.cs]: Fixed

File: src/VirtoCommerce.DemoSolutionFeaturesModule.Core/Models/Catalog/DemoProductPartItemInfo.cs (hunk 4 above)
Comment (asvishnyakov) [reply src/VirtoCommerce.DemoSolutionFeaturesModule.Core/Models/Catalog/DemoProductPartItemInfo.cs]: Fixed

File: src/VirtoCommerce.DemoSolutionFeaturesModule.Web/Controllers/Api/DemoCatalogController.cs (hunk 5 above)
Comment (asvishnyakov) [reply src/VirtoCommerce.DemoSolutionFeaturesModule.Web/Controllers/Api/DemoCatalogController.cs]: Fixed

**Review by pushnitsa — APPROVED** [review]

**Review by Vectorfield4 — APPROVED** [review]

**Comment by Vectorfield4** [comment]
Tests -> https://virtocommerce.atlassian.net/browse/VDS-748

**Review by pushnitsa — APPROVED** [review]

**Pull request merged**